File: ddd-application/src/main/java/victor/training/ddd/agile/entity/Sprint.java
package victor.training.ddd.agile.entity;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import victor.training.ddd.agile.service.EmailService;
import victor.training.ddd.agile.service.MailingListClient;

import javax.persistence.*;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import static javax.persistence.EnumType.STRING;

@Getter
@Setter
@NoArgsConstructor
@Entity
public class Sprint {
   @Id
   @GeneratedValue
   private Long id;
   private int iteration;
   @ManyToOne
   // TODO replace
   private Product product; // RULE: do not keep object links to other agg = Change contention
   // ==> a method in an aggragate can only be:
   // 1) a query method returining data
   // 2) change data inside = command
   // 2b) more side effects to other Agg / outside world (eg email send)    -> fire Event.
   // with
   private Long productId;
   private LocalDate startDate;
   private LocalDate plannedEndDate;
   private LocalDate endDate;

   public boolean allItemsAreDone() {
      return this.items.stream().allMatch(BacklogItem::isDone);
   }

   public void method() {
      product.setName("whatever");// PANIC moment; the callers of sprint.method()
      // do NOT expect the Product to change
   }

   public void startItem(long backlogId/*, BacklogItemRepo omg*/) {
      if (status != Status.STARTED) {
         throw new IllegalStateException();
      }
      itemById(backlogId).start();
   }


   public void complete(long backlogId,/*, BacklogItemRepo omg*/MailingListClient mailingListClient, EmailService emailService) {
      if (status != Status.STARTED) {
         throw new IllegalStateException();
      }
      itemById(backlogId).complete();

      // option 3: pass services as arg to methods of agg
      if (allItemsAreDone()) {
         System.out.println("Sending CONGRATS email to team of product " +
                            getProduct().getCode() + ": They finished the items earlier." +
                            " They have time to refactor! (OMG!)");
         List<String> emails = mailingListClient.retrieveEmails(getProduct().getTeamMailingList());
         emailService.sendCongratsEmail(emails);
      }
   }

   private BacklogItem itemById(long backlogId) {
      return items.stream().filter(i -> i.getId().equals(backlogId)).findFirst().orElseThrow();
   }

   public enum Status {
      CREATED,
      STARTED,
      FINISHED
   }

   @Enumerated(STRING)
   private Status status = Status.CREATED;

   @OneToMany(mappedBy = "sprint")
   private List<BacklogItem> items = new ArrayList<>();
}

File: ddd-application/src/main/java/victor/training/ddd/agile/entity/Sprint.kt
package victor.training.ddd.agile.entity

import org.springframework.context.ApplicationEventPublisher
import victor.training.ddd.agile.common.DomainEvents
import victor.training.ddd.agile.service.EmailService
import java.time.LocalDate
import javax.persistence.*

@Entity // DDD AggregateRoot

class Sprint(
    @ManyToOne
    val product: Product,
    val iteration:Int = 0,
    val plannedEndDate: LocalDate? = null,
    private var startDate: LocalDate? = null,
    private var endDate: LocalDate? = null,
    @Enumerated(EnumType.STRING)
    private var status:SprintStatus = SprintStatus.CREATED,
    @OneToMany(mappedBy = "sprint", fetch = FetchType.EAGER, cascade = [CascadeType.ALL])
    val items: MutableList<BacklogItem> = ArrayList(),
    @Id @GeneratedValue var id: Long? = null
) {

    enum class SprintStatus {
        CREATED, STARTED, FINISHED
    }

    fun end() {
        check(status == SprintStatus.STARTED)
        endDate = LocalDate.now()
        status = SprintStatus.FINISHED
    }

    fun start() {
        check(status == SprintStatus.CREATED)
        startDate = LocalDate.now()
        status = SprintStatus.STARTED
    }

    fun status() = status
    fun startDate() = startDate
    fun endDate() = endDate

    fun startItem(backlogId: Long) {
        check(status == SprintStatus.STARTED)
        itemById(backlogId).start()
    }

    private fun itemById(backlogId: Long) = items.stream()
        .filter { it.id == backlogId }
        .findFirst()
        .orElseThrow()

    fun completeItem(backlogId: Long, applicationEventPublisher: ApplicationEventPublisher) {
        check(status == SprintStatus.STARTED)
        itemById(backlogId).complete()
        if (items.all { it.isDone() }) {
            DomainEvents.publishEvent(SprintItemsFinishedEvent(id!!))
//            applicationEventPublisher.publishEvent(SprintItemsFinishedEvent(id!!))

            SprintItemsFinishedEvent(id!!)
        // passing the entire BL inside an Event is A BAD practice if the BL object is mutable !


//            emailService.sendCongratsEmail(product.code, product.teamMailingList) // couping me to the EmailService  / or a door of another module
        }
    }

    fun logHours(backlogId: Long, hours: Int) {
        check(status == SprintStatus.STARTED)
        itemById(backlogId).logHours(hours)
    }
}